include/shad/core/algorithm.h

```diff
@@ -401,222 +97,172 @@ ForwardItr find_if_not(ExecutionPolicy&& policy, ForwardItr first,
-          auto end = std::get<1>(args);
-          auto value = std::get<2>(args);
-
-          auto local_range = itr_traits::local_range(begin, end);
-          *result = std::count(local_range.begin(), local_range.end(), value);
-        },
-        std::make_tuple(first, last, value), &results[i]);
-
-    ++i;
-  }
-
-  rt::waitForCompletion(H);
-
-  return std::accumulate(results.begin(), results.end(), difference_type(0));
+    ExecutionPolicy&& policy, InputItr first, InputItr last, const T& value) {
+  return impl::count(std::forward<ExecutionPolicy>(policy), first, last, value);
 }
 
-template <typename InputItr, typename T>
-typename shad::distributed_iterator_traits<InputItr>::difference_type count(
-    distributed_sequential_tag&& policy, InputItr first, InputItr last,
-    const T& value) {
-  using itr_traits = distributed_iterator_traits<InputItr>;
-  using difference_type = typename itr_traits::difference_type;
-  auto localities = itr_traits::localities(first, last);
-
-  difference_type result = 0;
-  for (auto locality = localities.begin(), end = localities.end();
-       locality != end; ++locality) {
-    difference_type delta = 0;
-
-    rt::executeAtWithRet(
-        locality,
-        [](const std::tuple<InputItr, InputItr, T>& args,
-           difference_type* result) {
-          auto begin = std::get<0>(args);
-          auto end = std::get<1>(args);
-          auto value = std::get<2>(args);
-
-          auto local_range = itr_traits::local_range(begin, end);
-          *result = std::count(local_range.begin(), local_range.end(), value);
-        },
-        std::make_tuple(first, last, value), &delta);
-
-    result += delta;
-  }
-
-  return result;
+template <typename ExecutionPolicy, typename InputItr, typename UnaryPredicate>
+typename shad::distributed_iterator_traits<InputItr>::difference_type count_if(
+    ExecutionPolicy&& policy, InputItr first, InputItr last, UnaryPredicate p) {
+  return impl::count_if(std::forward<ExecutionPolicy>(policy), first, last, p);
 }
 
-}  // namespace impl
+// ---------------------------------------------//
+//                                              //
+//              minimum_maximum_ops             //
+//                                              //
+// ---------------------------------------------//
 
-template <typename ExecutionPolicy, typename InputItr, typename T>
-typename shad::distributed_iterator_traits<InputItr>::difference_type count(
-    ExecutionPolicy&& policy, InputItr first, InputItr last, const T& value) {
-  return impl::count(std::forward<ExecutionPolicy>(policy), first, last, value);
+//  -------------  //
+//  |   max     |  //
+//  -------------  //
+// Same as std::max  
+
+template <class T>
+const T& max(const T& a, const T& b) {
+  return std::max<T>(a, b);
 }
 
-namespace impl {
+template <class T, class Compare>
+const T& max(const T& a, const T& b, Compare comp) {
+  return std::max<T, Compare>(a, b, comp);
+}
 
```

Comment:
I think we should ignore the functions not accepting policies: we are not implementing a complete STL.

Reply:
I agree.  having a shad::max without the policy might induce the user to think that we are doing some smart thing under the hood while we are not.

Not including them, I feel we are more sincere :) with the api user in the sense that they know what they get.

Reply:
I think that unless there is a major issue with it, we should include all the methods, especially considering the zero-effort in maintaining them. This avoids the need of comparing the APIs, and trial-and-error (or make and pray) compilation.
Regarding Marco's point, in the header it is explicitly highlighted that those functions are "the same" as their stl counterpart. This will be specified also in the documentation, when we will add it.

Reply:
The major effort is having two different functions with the very same semantics, whereas a user is led to think that, if they are not the same function, they should do something different.
This would be needed if we were planning to provide a full STL implementation, but we are not, so I do not see any reason for including STL wrappers.
Finally, I do not see how this relates to trial-and-error scenarios.

Reply:
It may be true that having a full STL implementation is not a current goal; however, for the headers we are implementing, it would be better to have as much overlap as possible with the STL counterparts. Again, this is to avoid users familiar with the STL, to have to review the APIs to see what is and what is not included in the  SHAD "core" headers.
This may result in the trial-and-error scenario: users (reasonably) expecting all methods in a particular class (e.g. algorithm or numeric) to be implemented, may just use the shad namespace for such methods. And realize later that they are not there.

I'd also point out that the "same as" concept is widely adopted, also in the STL, and not just for overloads.

Reply:
I think we have good points on both sides.  What about taking a "defensive" position right now by not including them (just not to create expectations), and push the decision forward?

I feel that at this stage (as a user) I won't mind not having those methods and I would certainly prefer to have the extra stuff we are currently building.  What do you think?

This stuff is easy to fix when we will get to an agreement on the strategy to take.

Reply:
I agree, good points on both sides. I also agree with the defensive position: let us postpone this.